immer/single_hash_index.hpp

```diff
@@ -0,0 +1,318 @@
+#pragma once
+
+#include <immer/config.hpp>
+#include <immer/detail/hamts/champ.hpp>
+#include <immer/detail/hamts/champ_iterator.hpp>
+#include <immer/memory_policy.hpp>
+#include <type_traits>
+
+namespace immer {
+
+template <typename T,
+          typename KeyExtractor,
+          typename Hash,
+          typename Equal,
+          typename MemoryPolicy,
+          detail::hamts::bits_t B>
+class single_hash_index_transient;
+
+template <class T>
+struct default_key_extractor
+{
+    using key_type = decltype(T().id);
+
+    key_type operator()(const T& t) const { return t.id; }
+};
+
+template <typename T,
+          typename KeyExtractor = default_key_extractor<T>,
+          typename Hash         = std::hash<typename KeyExtractor::key_type>,
+          typename Equal = std::equal_to<typename KeyExtractor::key_type>,
+          typename MemoryPolicy   = default_memory_policy,
+          detail::hamts::bits_t B = default_bits>
+class single_hash_index
```

Comment:
I have been giving it some thought, and I think that `immer::table<>` would be a good name for this type. What do you think?

I don't like `single_hash_index`, it is very long and a bit confusing about what it means.